archive_repo.py

```diff
@@ -0,0 +1,667 @@
+    state
+    ...author
+    ...labels
+    ...commentFields
+    closedAt
+    comments(first: 100) {
+        nodes {
+            ...author
+            ...commentFields
+        }
+        pageInfo {
+            endCursor
+            hasNextPage
+        }
+        totalCount
+    }
+}
+"""
+    + gql_AuthorFields
+    + gql_Comment_Fields
+    + gql_LabelFields
+)
+
+gql_Issues_Query = (
+    """
+      nodes { ...issueFields }
+      pageInfo {
+        endCursor
+        hasNextPage
+      }
+      totalCount
+    }
+  }
+  rateLimit {
+    remaining
+    resetAt
+  }
+}
+"""
+    + gql_Issue_Fields
+)
+
+gql_AllIssues_First = (
+    """
+query($owner: String!, $repo: String!){
+  repository(owner: $owner, name: $repo) {
+    issues(first: 100) {
+"""
+    + gql_Issues_Query
+)
+
+gql_AllIssues_Subsequent = (
+    """
+query($owner: String!, $repo: String!, $cursor: String!){
+  repository(owner: $owner, name: $repo) {
+    issues(first: 100, after: $cursor) {
+"""
+    + gql_Issues_Query
+)
+
+gql_UpdatedIssues_First = (
+    """
+query($owner: String!, $repo: String!, $filters: IssueFilters){
+  repository(owner: $owner, name: $repo) {
+    issues(first: 100, filterBy: $filters) {
+"""
+    + gql_Issues_Query
+)
+
+gql_UpdatedIssues_Subsequent = (
+    """
+query($owner: String!, $repo: String!, $filters: IssueFilters, $cursor: String!){
+  repository(owner: $owner, name: $repo) {
+    issues(first: 100, filterBy: $filters, after: $cursor) {
+"""
+    + gql_Issues_Query
+)
+
+gql_Issue_Comments_Query = (
+    """
+query($id: ID!, $cursor: String!){
+    node(id: $id) {
+        ...on Issue {
+            comments(first:100, after:$cursor) {
+                nodes {
+                    ...author
+                    ...commentFields
+                }
+                pageInfo {
+                    endCursor
+                    hasNextPage
+                }
+                totalCount
+            }
+        }
+    }
+    rateLimit {
+        remaining
+        resetAt
+    }
+}
+"""
+    + gql_Comment_Fields
+    + gql_AuthorFields
+)
+
+# Pull Requests
+
+gql_Review_Fields = (
+    """
+fragment reviewFields on PullRequestReview {
+    id
+    commit { abbreviatedOid }
+    ...author
+    state
+    ...commentFields
+    comments(first: 50) {
+        nodes {
+            originalPosition
+            ...commentFields
+        }
+        pageInfo {
+            endCursor
+            hasNextPage
+        }
+        totalCount
+    }
+}
+"""
+    + gql_Comment_Fields
+    + gql_AuthorFields
+)
+
+gql_PullRequest_Fields = (
+    """
+fragment prFields on PullRequest {
+    number
+    id
+    title
+    url
+    state
+    ...labels
+    ...author
+    ...commentFields
+    closedAt
+    mergedAt
+    mergedBy { ...userFields }
+    comments(first: 100) {
+        nodes {
+            ...author
+            ...commentFields }
+        pageInfo {
+            endCursor
+            hasNextPage
+        }
+        totalCount
```

Comment:
It might pay to factor out this block into a gql_Paged variable.

Separately, I need to check that you really need totalCount for this.

Reply:
I tried, but couldn't get it to work as a fragment because fragments have to declare what type they're on, and these fields are perversely not part of any interface I could find in the documentation.  Python3 formatted strings choke on all the braces in a GraphQL expression, so it's not simple to just insert as an inline variable.  I could get it in with a sentinel and a `.replace()` call, but that doesn't really feel any cleaner than just repeating it.

If you figure out how, though, I'd be happy to take the change.

No, I don't use totalCount for anything; I was going to, but GitHub helpfully added the hasNextPage since I last took a stab at this.  While I thought there might be value in having it to cross-check, we can drop it with no loss of functionality (besides not having to count the JSON objects to display the statistic when traversing the file).

Reply:
This is just a string though, so you could do:

```python
gql_Paged = """
pageInfo {
    endCursor
    hasNextPage
}
"""

gql_PullRequest_Fields = (
    """
fragment prFields on PullRequest {
    number
    id
    title
    url
    state
    ...labels
    ...author
    ...commentFields
    closedAt
    mergedAt
    mergedBy { ...userFields }
    comments(first: 100) {
        nodes {
            ...author
            ...commentFields }
"""
+ gql_Paged 
"""
    }
    reviews(first: 50) {
        nodes {
            ...reviewFields
        }
"""
+ gql_Paged 
"""
    }
}
"""
    + gql_LabelFields
    + gql_Review_Fields
)
```

Or you could get fancy and do (with more formatting):

```python
def paged(name, fields, first=10, cursor=False):
    name + ("(first=%d" % first) + (", after:$cursor" if cursor else "") + ") { nodes { " + fields + "pageInfo { endCursor, hasNextPage } } }"
```

Reply:
I briefly sketched out something like the fancier approach, but there are too many different variations for me to think it's worthwhile, what with issue filters and parameter types.  By the time you pass in all the info it needs to construct the query, you're passing in the query.  The simpler version is there, though, with a "gql_Paged" between two block strings.